src/audio/smart_amp_maxim.c

```diff
@@ -0,0 +1,719 @@
+	default:
+		comp_err(dev, "smart_amp_process() error: not supported frame format");
+		return NULL;
+	}
+}
+
+static int smart_amp_copy(struct comp_dev *dev)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	uint32_t avail_passthrough_frames;
+	uint32_t avail_feedback_frames;
+	uint32_t avail_frames;
+	uint32_t source_bytes;
+	uint32_t sink_bytes;
+	uint32_t feedback_bytes;
+	uint32_t source_flags = 0;
+	uint32_t sink_flags = 0;
+	uint32_t feedback_flags = 0;
+	int ret = 0;
+
+	comp_dbg(dev, "smart_amp_copy()");
+
+	buffer_lock(sad->source_buf, &source_flags);
+	buffer_lock(sad->sink_buf, &sink_flags);
+
+	/* available bytes and samples calculation */
+	avail_passthrough_frames =
+		audio_stream_avail_frames(&sad->source_buf->stream,
+					  &sad->sink_buf->stream);
+
+	buffer_unlock(sad->source_buf, source_flags);
+	buffer_unlock(sad->sink_buf, sink_flags);
+
+	avail_frames = avail_passthrough_frames;
+
+	buffer_lock(sad->feedback_buf, &feedback_flags);
+	if (sad->feedback_buf->source->state == dev->state) {
+		/* feedback */
+		avail_feedback_frames = sad->feedback_buf->stream.avail /
+			audio_stream_frame_bytes(&sad->feedback_buf->stream);
+
+		avail_frames = MIN(avail_passthrough_frames,
+				   avail_feedback_frames);
+
+		feedback_bytes = avail_frames *
+			audio_stream_frame_bytes(&sad->feedback_buf->stream);
+
+		buffer_unlock(sad->feedback_buf, feedback_flags);
+
+		comp_dbg(dev, "smart_amp_copy(): processing %d feedback frames (avail_passthrough_frames: %d)",
+			 avail_frames, avail_passthrough_frames);
+
+		sad->process(dev, &sad->feedback_buf->stream,
+			     &sad->sink_buf->stream, avail_frames,
+			     sad->config.feedback_ch_map, true);
+
+		comp_update_buffer_consume(sad->feedback_buf, feedback_bytes);
+	}
+
+	/* bytes calculation */
+	buffer_lock(sad->source_buf, &source_flags);
+	source_bytes = avail_frames *
+		audio_stream_frame_bytes(&sad->source_buf->stream);
+	buffer_unlock(sad->source_buf, source_flags);
+
+	buffer_lock(sad->sink_buf, &sink_flags);
+	sink_bytes = avail_frames *
+		audio_stream_frame_bytes(&sad->sink_buf->stream);
+	buffer_unlock(sad->sink_buf, sink_flags);
+
+	/* process data */
+	sad->process(dev, &sad->source_buf->stream, &sad->sink_buf->stream,
+		     avail_frames, sad->config.source_ch_map, false);
+
+	/* source/sink buffer pointers update */
+	comp_update_buffer_consume(sad->source_buf, source_bytes);
+	comp_update_buffer_produce(sad->sink_buf, sink_bytes);
+
+	return ret;
+}
+
+static int smart_amp_reset(struct comp_dev *dev)
+{
+	comp_dbg(dev, "smart_amp_reset()");
+
+	comp_set_state(dev, COMP_TRIGGER_RESET);
+
+	return 0;
+}
+
+static int smart_amp_prepare(struct comp_dev *dev)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct comp_buffer *source_buffer;
+	struct list_item *blist;
+	int ret;
+
+	comp_dbg(dev, "smart_amp_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
```

Comment:
ditto, are we setting same state twice ?

Reply:
Same here. Removed this condition check.